### PR TITLE
Iterative

### PR DESCRIPTION
**Pull request by tterraz**
Implementation of iterative algorithms.

Available: mean, variance, skewness, kurtosis, min, max, threshold exceedance, Sobol' indices and quantiles. Formulas still need to be checked, and documentation to be completed.

### CODE REVIEW

**Comment by mbaudin47** [comment]
My own build produces a set of errors:

/home/osboxes/GIT/openturns-iterative/openturns/build/python/src/sphinx_build/examples/iterative_statistics/iterative_statistics.rst:4: WARNING: toctree contains reference to document 'examples/iterative_statistics/iterative_variance' that doesn't have a title: no link will be generated
/home/osboxes/GIT/openturns-iterative/openturns/build/python/src/sphinx_build/examples/iterative_statistics/iterative_statistics.rst:4: WARNING: toctree contains reference to document 'examples/iterative_statistics/iterative_sobol_jansen' that doesn't have a title: no link will be generated
/home/osboxes/GIT/openturns-iterative/openturns/build/python/src/sphinx_build/examples/iterative_statistics/iterative_statistics.rst:4: WARNING: toctree contains reference to document 'examples/iterative_statistics/iterative_sobol_martinez' that doesn't have a title: no link will be generated
/home/osboxes/GIT/openturns-iterative/openturns/build/python/src/sphinx_build/examples/iterative_statistics/iterative_statistics.rst:4: WARNING: toctree contains reference to document 'examples/iterative_statistics/iterative_sobol_mauntz_kucherenko' that doesn't have a title: no link will be generated
/home/osboxes/GIT/openturns-iterative/openturns/build/python/src/sphinx_build/examples/iterative_statistics/iterative_statistics.rst:4: WARNING: toctree contains reference to document 'examples/iterative_statistics/iterative_sobol_saltelli' that doesn't have a title: no link will be generated

**Comment by mbaudin47** [comment]
According to appveyor, the following test are failing.
    pyinstallcheck_SobolIndicesAlgorithm_std
    pyinstallcheck_IterativeSkewness_std
    pyinstallcheck_IterativeMean_std

One of the errors is:
```
    Traceback (most recent call last):
  File "C:/projects/openturns/python/test/t_IterativeMean_std.py", line 5, in <module>
    import numpy.testing as nptest
ImportError: No module named numpy.testing
```

**Comment by mbaudin47** [comment]
There seem to be unused variables in the C++ code. Please look at the end of:
https://circleci.com/api/v1.1/project/github/openturns/openturns/5401/output/103/0?file=true&allocation-id=5e81f480067b182ec08be56d-0-build%2F258A521C

File: lib/src/Base/IterativeStat/openturns/IterativeAlgorithmImplementation.hxx [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+   * Default constructor
+   *
+   * The constructor sets a new Id to the object,
+   * so it can be later referenced by a Study object.
+   * It is also declared visible if member of a study.
+   *
+   * The object has the default name but it does not
+   * use storage for it.
+   */
+//   IterativeAlgorithmImplementation()
```
Comment (mbaudin47) [inline lib/src/Base/IterativeStat/openturns/IterativeAlgorithmImplementation.hxx:31]: Please remove commented code.

File: lib/src/Base/IterativeStat/openturns/IterativeAlgorithmImplementation.hxx [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+   * Increment methods
+   *
+   * @warning These methods MUST be overloaded in derived classes.
+   */
+
+  virtual void increment(const Point & newData) = 0;
+
+  virtual void increment(const Sample & newData) = 0;
+
+//   virtual void finalize()
```
Comment (mbaudin47) [inline lib/src/Base/IterativeStat/openturns/IterativeAlgorithmImplementation.hxx:69]: Please remove commented code.

File: lib/src/Base/IterativeStat/openturns/IterativeAlgorithmImplementation.hxx [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+//   {
+//     // Nothing to do
+//   }
+
+  /**
+   * Iteration accessor
+   *
+   * This method returns the current iteration number of the algorithm.
+   */
+  UnsignedInteger getIteration() const
```
Comment (mbaudin47) [inline lib/src/Base/IterativeStat/openturns/IterativeAlgorithmImplementation.hxx:79]: The code must be in the cxx, not in the hxx.

File: lib/src/Base/IterativeStat/openturns/IterativeAlgorithmImplementation.hxx [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+   *
+   * This method compares objects based on their content.
+   */
+  inline virtual
+  Bool operator !=(const IterativeAlgorithmImplementation & other) const
+  {
+    return !operator==(other);
+  }
+
+  /* String converter */
```
Comment (mbaudin47) [inline lib/src/Base/IterativeStat/openturns/IterativeAlgorithmImplementation.hxx:116]: The code must be in the cxx, not in the hxx. Please look at https://github.com/openturns/openturns/blob/master/lib/src/Base/Func/openturns/BasisSequenceImplementation.hxx for an example. This generates an error at build time.

File: lib/src/Uncertainty/Algorithm/Sensitivity/MartinezSobolIndices.cxx [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+    covarianceI_ = Sample(modelInputDimension_, modelOutputDimension_);
+    covarianceTI_ = Sample(modelInputDimension_, modelOutputDimension_);
+    if (iteration_ != 0)
+      LOGWARN(OSS() << "Already computed indices, previous values will be lost");
+    iteration_ = 0;
+  }
+
+  const UnsignedInteger size = inputSample.getSize() / (modelInputDimension_ + 2);
+
+//   Sample yA(inputSample, 0, size);
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Algorithm/Sensitivity/MartinezSobolIndices.cxx:150]: Please remove commented code.

File: python/doc/examples/data_analysis/data_analysis.rst [hunk 6]
```diff
@@ -81,6 +81,7 @@ Estimate stochastic processes
     estimate_non_stationary_covariance_model.ipynb
     estimate_spectral_density_function.ipynb
 
+<<<<<<< HEAD
```
Comment (mbaudin47) [inline python/doc/examples/data_analysis/data_analysis.rst:84]: This is a failed merged. Please fix the rst.

File: python/doc/examples/data_analysis/data_analysis.rst [hunk 7]
```diff
@@ -90,3 +91,5 @@ Graphics
     visualize_clouds.ipynb
     visualize_pairs.ipynb
 
+=======
```
Comment (mbaudin47) [inline python/doc/examples/data_analysis/data_analysis.rst:94]: This is a failed merged. Please fix the rst.

File: python/doc/examples/iterative_statistics/README.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+## Table of Contents
+* Overview
+* Iterative Statistics
+* Iterative Sensitivity Analysis
+
+## Overview
+
+OpenTURNS provides a lot of algorithms to compute classical statistics and Sobol indices. However, these algorithms are not designed to compute such statistics in an iterative fashion. In OpenTURNS, values of interest are computed directly from a sample or an experiment design. From a complete set of data, we compute the value of interest. This aproach can no longer scale with the size of the data set.
+Iterative Algorithms can increment statistics with a subset of the data, and then are no longer limited by the size of the data set itself. They require a standalone data structure that can be fed with the samples iteratively. This point of view is not classical in OpenTurns. The data structure of the value of interest must persist and be updated when given new data. In this document, we present the  work of Inria to implemente iterative algorithms in OpenTURSN.
```
Comment (mbaudin47) [inline python/doc/examples/iterative_statistics/README.md:14]: ~~OpenTURSN~~ OpenTURNS

File: python/doc/examples/iterative_statistics/README.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+* IterativeThreshold exceedance
+* Mean
+* Variance
+* Skewness
+* Kurtosis
+* Quantiles (not validated)
+
+These objects are instanciable. They all inherit the getIteration() and getDimension() methods from the IterativeAlgorithm class, and overload the Increment(Point) and Increment(Sample) methods. These objects are standalone, so it is the responsibility of the user to manage the experiments that will generate the samples for the iterative algorithms.
+
+These classes do not keep track of the input samples, but some values of interest can still be derived from the main statistic.
```
Comment (mbaudin47) [inline python/doc/examples/iterative_statistics/README.md:37]: ~~values~~ of interest > ~~quantity~~ of interest

File: python/doc/examples/iterative_statistics/README.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+For example, the mean can be extracted from the internal data structure of the iterative variance, so the IterativeVariance class has a getMean() method.
+
+All these classical iterative statistics do not affect the other parts of OpenTURNS.
+
+
+## Iterative Sensitivity Analysis
+
+As for the classical statistics, OpenTURNS is not designed for itertaive sensitivity analysis. So as for the classical statistics, we defined a persistant object that contains the data structure for the Sobol' indices and the algorithms to compute them. This object is a SobolIndices object, its implementation is in  lib/src/Uncertainty/Algorithm/Sensitivity. It is similar to Iterative Algorithm: it inherits from persistentObject and is designed with a bridge patern. The 4 classes that inherit from this class are: JansenSobolIndices, MartinezSobolIndices, MauntzKucherenkoSobolIndices and SaltelliSobolIndices. These classes implement the methods to compute the corresponding Sobol' indices. In order to be consistent with the SobolIndicesAlgorithm classes, we mooved the classical algorithm from
```
Comment (mbaudin47) [inline python/doc/examples/iterative_statistics/README.md:45]: ~~itertaive~~ > iterative

Comment (mbaudin47) [reply python/doc/examples/iterative_statistics/README.md:45]: ~~waay~~ > way

**Comment by jschueller** [comment]
This PR is rotting: no news lot of work is needed to catch up to master. It can still be resubmitted if needed.

**Pull request closed without merging**